Add remaining array FFI functions

sharno opened the pull request:
I think this completes the arrays.Array functionality

sharno committed:
Add `replicate` and `formFoldableImpl`
Add sections similar to the javascript ffi file

sharno committed:
Add all remaining array ffi functions

sharno committed:
Fix insertAt edge cases

andyarvanitis commented:
Thanks, I'll review this weekend (I ran out of time tonight). Two quick requests:

1) I've recently added some coding conventions to the [README](https://github.com/andyarvanitis/purescript-native-go-ffi/blob/master/README.md). Please make sure to follow them as best you can (and feel free to ask me any questions). 

2) If you're completing the FFI for a purescript package (in this case `purescript-arrays`), can you run its test suite (under `purescript-arrays/test`) to see if everything passes? If anything it depends on is missing or broken, you can just let me know.

sharno commented:
Take your time, I was planning definitely to run the tests, I'm not sure about the configuration though to try it out but I'll see how it goes

I tried to follow the conventions as close as possible, I'll make sure that `go fmt` is run but I think that vscode ran it automatically for me when I saved.
One question though, for the convention of catching the error and throwing it away for Type Assertion, I tried to benchmark it locally and it doesn't seem to have any effect performance, and I think it's better to just leave the ability to panic in case something was wrong so the user would know how to investigate or report a bug?

sharno committed:
- Fix a bug in findIndex of Array
- Add more FFI functions needed for the Array package tests to run

sharno commented:
I pushed some more functions in ST and other places because the tests in the arrays package need them
I have a question though, I couldn't implement `pushAll` because it needs the pointer of the slice to be passed in not a copy of it. So what do you think could be a solution?
My only thought now is we'll need to change the whole implementation of `Array` to be a struct having the an `Array` reference?

andyarvanitis commented:
> I tried to follow the conventions as close as possible, I'll make sure that go fmt is run but I think that vscode ran it automatically for me when I saved.

Yup, I'm guessing it does it, as do some other editors with Golang support.

> One question though, for the convention of catching the error and throwing it away for Type Assertion, I tried to benchmark it locally and it doesn't seem to have any effect performance...

I think the performance difference is kind of a "YMMV" deal. Granted, the last time I read about it in any detail was a while ago, so maybe the overhead is now the same for either approach, and certainly the panics are much more user-friendly. In fact, I'm just about to check in some runtime changes for a debug build mode with the panics enabled. I think I'll keep the guidance as-is for now, but I will dig into further to see if I should revise it. In the meantime, I'm fine with it either way if you're already using the panicking version, just please try to stick to the "`:=`" notation for variable creation.

> I pushed some more functions in ST and other places because the tests in the arrays package need them. I have a question though, I couldn't implement `pushAll` because it needs the pointer of the slice to be passed in not a copy of it. So what do you think could be a solution?
My only thought now is we'll need to change the whole implementation of `Array` to be a struct having the an `Array` reference?

The representation of immutable arrays shouldn't be changed, but `ST` foreign functions should deal with pointers. So for example, in that module, `empty` should be implemented as:
```go
func init() {
	exports := Foreign("Data.Array.ST")
	
	exports["empty"] = func() Any {
		e := make([]Any, 0)
		return &e
	}
...
 ```
or maybe more simply:
```go
	exports["empty"] = func() Any {
		return new([]Any)
	}
```


Also see [purescript-refs ffi](https://github.com/andyarvanitis/purescript-native-go-ffi/blob/master/purescript-refs/Effect_Ref.go) for similar ideas/examples

sharno commented:
>The representation of immutable arrays shouldn't be changed, but ST foreign functions should deal with pointers

I think the problem is in the purescript implementations assuming that an array is always a reference in JS, but it's not in Go
So if you look at those functions for `STArray` in `purescript-arrays`:

```purescript
-- | O(1). Convert a mutable array to an immutable array, without copying. The mutable
-- | array must not be mutated afterwards.
unsafeFreeze :: forall h a. STArray h a -> ST h (Array a)
unsafeFreeze = pure <<< (unsafeCoerce :: STArray h a -> Array a)

-- | O(1) Convert an immutable array to a mutable array, without copying. The input
-- | array must not be used afterward.
unsafeThaw :: forall h a. Array a -> ST h (STArray h a)
unsafeThaw = pure <<< (unsafeCoerce :: Array a -> STArray h a)

nubBy :: forall a. (a -> a -> Ordering) -> Array a -> Array a
nubBy comp xs = case head indexedAndSorted of
  Nothing -> []
  Just x -> map snd $ sortWith fst $ ST.run do
     -- TODO: use NonEmptyArrays here to avoid partial functions
     result <- STA.unsafeThaw $ singleton x
     ST.foreach indexedAndSorted \pair@(Tuple i x') -> do
       lst <- snd <<< unsafePartial (fromJust <<< last) <$> STA.unsafeFreeze result
       when (comp lst x' /= EQ) $ void $ STA.push pair result
     STA.unsafeFreeze result
  where
  indexedAndSorted :: Array (Tuple Int a)
  indexedAndSorted = sortBy (\x y -> comp (snd x) (snd y))
                            (mapWithIndex Tuple xs)
```


I think what happens in `nubBy` is the slice passed as value from singleton (it returns a slice not a pointer) to unsafeThaw. This causes everything else that depends on referenced to be passed by value. It was written probably assuming that JS would always deal with references not values like in Go.
I think this is not a problem with Refs because everything is dealt with from the beginning as Go pointers.
Unless we deal with everything in arrays as a pointer?

andyarvanitis commented:
Ok, thanks, I see what you mean. They are assuming and using `unsafeCoerce`. I'll give it some thought. One approach might be a PR to the purescript guys to remove that assumption and change things a bit (like making `unsafeFreeze` and `unsafeThaw` foreign), but let me look at it more before going there.

sharno commented:
Another example is `copyImpl` which is used in both `thaw` and `freeze`, again if we need `STArray` to be a pointer they'll need to have different FFI implementations:
```purescript
-- | Create a mutable copy of an immutable array.
thaw :: forall h a. Array a -> ST h (STArray h a)
thaw = copyImpl

-- | Create an immutable copy of a mutable array.
freeze :: forall h a. STArray h a -> ST h (Array a)
freeze = copyImpl
```

andyarvanitis commented on purescript-arrays/Data_Array_ST.go:
```diff
@@ -0,0 +1,33 @@
+				// TODO: buggy implementation, we need to update the reference not create a new slice
+				as := (as_).([]Any)
+				xs := xs_.([]Any)
+				as = append(as, xs...)
```
This alone won't fix it (due to the reference issue we've been discussing), but I believe this should be
```go
xs = append(xs, as...)
return len(xs)
```

andyarvanitis commented:
I've looked at it some more, and am pretty sure a PR with local replacements for the use of `unsafeCoerce` are in order. I'll start putting something together. In the meantime, you can hack `unsafeCoerce` itself (combined with using `*[]Any` for `ST` arrays in these foreign functions) in the following way:

Edit: scratch that, apparently I'm too tired to post anything useful right now

Edit edit: ok, I'm still tired, but here it is, since I think it should work fine, at least for just testing:

```go
package purescript_unsafe_coerce

import (
	. "github.com/purescript-native/go-runtime"
)

func init() {
	exports := Foreign("Unsafe.Coerce")

	exports["unsafeCoerce"] = func(x_ Any) Any {
		switch x := x_.(type) {
		case []Any:
			return &x
		case *[]Any:
			return *x
		default:
			return x
		}
	}
}
```

And just to be clear, `pushAll` can then be:
```go
	as := (as_).([]Any)
	xs := xs_.(*[]Any)
	*xs = append(*xs, as...)
	return len(*xs)
```

andyarvanitis commented on purescript-arrays/Data_Array_ST.go:
```diff
@@ -0,0 +1,33 @@
+				as := (as_).([]Any)
+				xs := xs_.([]Any)
+				as = append(as, xs...)
+				return len(as)
```
and `return len(xs)`

sharno committed:
- Fix the STArray representation
- Temporary solution for unsafeCoerce to deal with STArray

sharno commented:
I applied the fixes you posted. But there's something weird happening that doesn't make all tests pass. It seems that `unsafeFreeze/unsafeThaw` doesn't call `unsafeCoerce` for some reason. I tried to check the generated code it's not there.

Not sure if this is a bug in this code or is it a compiler thing

sharno committed:
Fix missing default case in copyImpl

andyarvanitis commented:
I just remembered that `unsafeCoerce` gets inlined by the compiler: https://github.com/andyarvanitis/purescript-native/blob/golang/src/CodeGen/IL/Optimizer.hs

If you're currently building `psgo` yourself, you can remove it and try it again.

I only meant to suggest the `unsafeCoerce` hack for testing your changes, not as a real solution anyway (or as a stop-gap until `purescript-arrays` could get re-worked)

sharno committed:
Add `peekImpl` FFI function to STArray

sharno commented:
Ok, removing the `unsafeCoerce` inlining and adding one more FFI function, all tests from the Array (just Array not STArray or NonEmpty) passed
I think this is enough for this PR, I can open different PRs for other functions in ST and NonEmpty

andyarvanitis commented:
PR created: https://github.com/purescript/purescript-arrays/pull/159

We would then just need:
```go
	exports["unsafeFreeze"] = func(xs_ Any) Any {
		return func() Any {
			xs, _ := xs_.(*[]Any)
			return *xs
		}
	}

	exports["unsafeThaw"] = func(xs_ Any) Any {
		return func() Any {
			xs, _ := xs_.([]Any)
			return &xs
		}
	}

	exports["freeze"] = func(xs_ Any) Any {
		return func() Any {
			xs, _ := xs_.(*[]Any)
			return append([]Any{}, *xs...)
		}
	}

	exports["thaw"] = func(xs_ Any) Any {
		return func() Any {
			xs, _ := xs_.([]Any)
			result := append([]Any{}, xs...)
			return &result
		}
	}
```

andyarvanitis commented:
@sharno  Now that the `purescript-arrays` PR has been merged, were you planning on updating this PR, or should I merge what you have and then make changes? (including reverting the `unsafeCoerce` change). I'm not in a big hurry, just wondering

sharno committed:
add new foreign functions freeze, thaw, unsafeFreeze, unsafeThaw
remove copyImpl foreign function
revert unsafeCoerce to just return its arguments

sharno commented:
@andyarvanitis fixed

Though there are a few tests in the Array module that don't pass, those that use these functions in the `Data.Array.NonEmpty` module, specifically the ones that use `unsafeCoerce`

```
unsafeFromArray :: forall a. Array a -> NonEmptyArray a
unsafeFromArray = unsafeCoerce

unsafeFromArrayF :: forall f a. f (Array a) -> f (NonEmptyArray a)
unsafeFromArrayF = unsafeCoerce

toArray :: forall a. NonEmptyArray a -> Array a
toArray = unsafeCoerce
```

They just throw an error:
```
panic: Foreign value 'unsafeCoerce' for purescript module 'Unsafe.Coerce' not found
```

Is that because of the inlining of unsafeCoerce?
I didn't build a version of the compiler that doesn't inline it to see if it's the reason as I changed my machine

Otherwise, I guess now you can merge this PR and we can look into that issue separately

andyarvanitis commented:
I've merged your changes, thanks again. I'm not seeing a problem with `unsafeCoerce` – maybe make sure everything in your environment is clean and up-to-date? The missing foreign value I see for the NonEmpty tests (`testNonEmptyArray`) is `unfoldr1ArrayImpl`
